spec/inputs/dynamodb_spec.rb

```diff
@@ -0,0 +1,207 @@
+# encoding: utf-8
+#
+#Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+#Licensed under the Apache License, Version 2.0 (the "License");
+#you may not use this file except in compliance with the License.
+#You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#Unless required by applicable law or agreed to in writing, software
+#distributed under the License is distributed on an "AS IS" BASIS,
+#WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#See the License for the specific language governing permissions and
+#limitations under the License.
+#
+require "logstash/devutils/rspec/spec_helper"
+require "logstash/inputs/dynamodb"
+require "rspec/expectations"
+require "rspec/mocks"
+require "mocha"
+
+RSpec.configure do |config|
+  config.mock_with :mocha
+end
+
+class LogStash::Inputs::TestDynamoDB < LogStash::Inputs::DynamoDB
+  default :codec, 'json'
+
+  private
+  def shutdown_count
+    @shutdown_count ||= 0
+  end
+
+  def queue_event(event, logstash_queue, host)
+    super(event, logstash_queue, host)
+    # Add additional item to plugin's queue to ensure run() flushes queue before shutting down.
+    # Queue the event and then shutdown, otherwise the threads would run forever
+    if shutdown_count === 0
+      @shutdown_count += 1
+      @queue << "additional event stuck in queue during shutdown"
+      raise LogStash::ShutdownSignal
+    end
+  end
+
+  def start_kcl_thread()
+    @queue << "some message from kcl thread calling process"
+  end
+end
+
+class TestParser
+
+  def parse_scan(msg)
+    return msg
+  end
+
+  def parse_stream(msg)
+    return msg
+  end
+
+end
+
+describe 'inputs/dynamodb' do
+  let (:empty_config) {{}}
+  let (:tablename) {{'table_name' => 'test tablename', 'view_type' => "new_and_old_images", "endpoint" => "some endpoint"}}
+  let (:invalid_aws_credentials_config) {{'table_name' => 'test tablename', "endpoint" => "some endpoint", 'aws_access_key_id' => 'invalid', 'aws_secret_access_key' => 'invalid_also', 'view_type' => "new_and_old_images", "streams_endpoint" => "some streams endpoint"}}
+  let (:invalid_aws_credentials_config_no_endpoints) {{'table_name' => 'test tablename', 'aws_access_key_id' => 'invalid', 'aws_secret_access_key' => 'invalid_also', 'view_type' => "new_and_old_images"}}
+  let (:dynamodb_client) {mock("AmazonDynamoDB::AmazonDynamoDBClient")}
+  let (:dynamodb_streams_client) {mock("AmazonDynamoDB::AmazonDynamoDBStreamsClient")}
+  let (:adapter) {mock("AmazonDynamoDB::AmazonDynamoDBStreamsAdapterClient")}
+  let (:parser) {mock("DynamoDBLogParser")}
+  let (:region_utils) {mock("RegionUtils")}
+
+  def allow_invalid_credentials(stream_status = "ENABLED", error_to_raise = nil)
```

Comment:
I would also recommend moving this into a helper, more information on helpers for rspec could be found at https://www.relishapp.com/rspec/rspec-core/docs/helper-methods/define-helper-methods-in-a-module